driver-core/src/main/com/mongodb/connection/netty/NettyStream.java

```diff
@@ -152,13 +199,20 @@ public void initChannel(final SocketChannel ch) {
                             enableHostNameVerification(sslParameters);
                         }
                         engine.setSSLParameters(sslParameters);
-                        ch.pipeline().addFirst("ssl", new SslHandler(engine, false));
+                        pipeline.addFirst("ssl", new SslHandler(engine, false));
                     }
+
                     int readTimeout = settings.getReadTimeout(MILLISECONDS);
-                    if (readTimeout > 0) {
-                        ch.pipeline().addLast(READ_HANDLER_NAME, new ReadTimeoutHandler(readTimeout));
+                    if (readTimeout > NO_SCHEDULE_TIME) {
+                        readTimeoutMillis = readTimeout;
+                        /* We need at least one handler before (in the inbound evaluation order) the InboundBufferHandler,
```

Comment:
I don't really understand this comment.  Where are we firing exception events?  What would be the observed behavior if we didn't have this handler?

Reply:
> Where are we firing exception events?
[`ctx.fireExceptionCaught(ReadTimeoutException.INSTANCE)`](https://github.com/mongodb/mongo-java-driver/pull/636/files#diff-38fbf22cab5e77e01fe9fc24165ab0979e93cc5681d7b139798686145f1744b9R552) fires it.

The only way for us to fire an exception event in Netty not as a result of handing an inbound/outbound event in [`ChannelInboundHandler`](https://netty.io/4.1/api/io/netty/channel/ChannelInboundHandler.html)/[`ChannelOutboundHandler`](https://netty.io/4.1/api/io/netty/channel/ChannelOutboundHandler.html), but on demand is by calling [`ChannelHandlerContext.fireExceptionCaught`](https://netty.io/4.1/api/io/netty/channel/ChannelHandlerContext.html#fireExceptionCaught-java.lang.Throwable-). This method results [_"in having the ChannelInboundHandler.exceptionCaught(ChannelHandlerContext, Throwable) method called of the next ChannelInboundHandler contained in the ChannelPipeline of the Channel"_](https://netty.io/4.1/api/io/netty/channel/ChannelHandlerContext.html#fireExceptionCaught-java.lang.Throwable-). So we need to store a reference to a context of a handler (we store it in `ReadTimeoutTask.ctx`), which is a legit approach: [_"You can keep the ChannelHandlerContext for later use, such as triggering an event outside the handler methods, even from a different thread."_](https://netty.io/4.1/api/io/netty/channel/ChannelHandlerContext.html).

Now we need to decide which handler's context we will use. Our `InboundBufferHandler` knows how to process exceptions, so we need timeout exception events to be forwarded to this handler, which means we cannot use its context for firing such events (this follows from the docs quoted above). 👉 At this point you may read the [comment](https://github.com/mongodb/mongo-java-driver/pull/636/files#diff-38fbf22cab5e77e01fe9fc24165ab0979e93cc5681d7b139798686145f1744b9R208) again, and it will probably be clear this time. Hope this helps, but if not, let's have a call.